brcm63xx: Add support for D-Link DSL-2750u rev C1

naseef opened the pull request:
This adds support for the D-Link DSL-2750u rev C1.
(https://wikidevi.com/wiki/D-Link_DSL-2750U_rev_C1)

It uses the same hardware as ADB P.DG A4001N.
CPU:   Broadcom BCM63281 (320 MHz)
RAM:   324M
FLASH: 8M

Flash instructions:

1.  Assign static IP 192.168.1.100 to PC
2.  Unplug the power source
3.  Press the RESET button at the router, don't release it yet!
4.  Plug the power source.Wait some seconds
5.  Release the RESET button
6.  Browse to http://192.168.1.1
7.  Send the openwrt-brcm63xx-generic-DSL2750U-C1-squashfs-cfe.bin and
    wait some minutes until the firmware upgrade finish.

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>

ynezz commented on target/linux/brcm63xx/dts/dsl-2750u-c1.dts:
```diff
@@ -0,0 +1,101 @@
+/dts-v1/;
```
@naseef Please can you consider `SPDX-License-Identifier: GPL-2.0-or-later OR MIT` as in [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist) ?
> Having MIT as alternative is sometimes preferred by upstream maintainers
and allows sharing that simple code with other projects. We don't really
want multiple DTS versions for the same device.

ynezz commented on target/linux/brcm63xx/dts/dsl-2750u-c1.dts:
```diff
@@ -0,0 +1,102 @@
+		};
+	};
+
+	gpio-leds {
```
@naseef  Could you please change this to generic name `leds` as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
>  5. The name of a node should reflect the function of the device and not its model. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation

While at it, could you please add newline between LED nodes:

```
                inet_red {
			label = "DSL2750U-C1:red:inet";
			gpios = <&pinctrl 1 0>;
		};

		power_red {
			label = "DSL2750U-C1:red:power";
			gpios = <&pinctrl 4 0>;
		};
```

```

ynezz replied:
And probably switch to lowercase board name in LED label `dsl-2750u-c1:red:power`.

ynezz commented on target/linux/brcm63xx/dts/dsl-2750u-c1.dts:
```diff
@@ -0,0 +1,102 @@
+		stdout-path = "serial0:115200n8";
+	};
+
+	gpio-keys-polled {
```
@naseef  Could you please change this to generic name `keys` as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
>  5. The name of a node should reflect the function of the device and not its model. Examples for generic node names can be found in Section 2.2.2 Generic Names Recommendation

ynezz commented on target/linux/brcm63xx/dts/dsl-2750u-c1.dts:
```diff
@@ -0,0 +1,102 @@
+
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
```
@naseef Drop `#address-cells = <1>;` and `#size-cells = <0>;` as it's not necessary

ynezz commented on target/linux/brcm63xx/dts/dsl-2750u-c1.dts:
```diff
@@ -0,0 +1,102 @@
+/ {
+	model = "D-Link DSL-2750u rev C1";
+	compatible = "d-link,dsl-2750u-c1", "brcm,bcm6328";
+
```
@naseef  Could you please add LEDs aliases as suggested by [DTS Checklist](https://openwrt.org/submitting-patches#dts_checklist)
> 4. If it's possible try to dedicate some of the LEDs for system status indication in [diag.sh](https://github.com/openwrt/openwrt/blob/master/package/base-files/files/etc/diag.sh) ([example](https://github.com/openwrt/openwrt/blob/master/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dts#L13))
https://github.com/openwrt/openwrt/blob/fe7d965ea95e78905328fe5425c8e90e3bf11e58/target/linux/ath79/dts/qca9533_tplink_tl-wr841-v11.dts#L13-L18

ynezz replied:
Well it seems like is still not using generic diag.sh, so you can probably skip this one.

naseef commented:
Suggestions incorporated.

naseef committed:
brcm63xx: Add support for D-Link DSL-2750u rev C1

This adds support for the D-Link DSL-2750u rev C1.
(https://wikidevi.com/wiki/D-Link_DSL-2750U_rev_C1)

It uses the same hardware as ADB P.DG A4001N.
CPU:   Broadcom BCM63281 (320 MHz)
RAM:   324M
FLASH: 8M

Flash instructions:

1.  Assign static IP 192.168.1.100 to PC
2.  Unplug the power source
3.  Press the RESET button at the router, don't release it yet!
4.  Plug the power source.Wait some seconds
5.  Release the RESET button
6.  Browse to http://192.168.1.1
7.  Send the openwrt-brcm63xx-generic-DSL2750U-C1-squashfs-cfe.bin and
    wait some minutes until the firmware upgrade finish.

Signed-off-by: Ahmed Naseef <naseefkm@gmail.com>

naseef commented:
@KanjiMoster any review on this PR?

naseef commented:
@KanjiMonster Can you please review this PR?

mwarning commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -549,6 +549,19 @@ define Device/DSL274XB-F1
+define Device/DSL2750U-C1
+  $(Device/bcm63xx)
+  IMAGES += sysupgrade.bin
+  DEVICE_TITLE := D-Link DSL-2750U rev C1
```
Please split up DEVICE_TITLE into
```
DEVICE_VENDOR := D-Link
DEVICE_MODEL := DSL-2750U
DEVICE_VARIANT := C1
```
(not sure about about "rev C1" vs. "C1" here)

See https://github.com/openwrt/openwrt/commit/8c3dbaf8b49233e275b2dce4e30dc0a2bf8c86c5

adrianschmutzler commented:
I've just pushed a tidy-up/reorganization patch for brcm63xx target.
For you, this would mean that you have to update device name, DTS name and the board name in board-files. I will comment inline to make this more clear...

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -549,6 +549,19 @@ define Device/DSL274XB-F1
 endef
 TARGET_DEVICES += DSL274XB-F1
 
+define Device/DSL2750U-C1
```
Device/d-link_dsl-2750u-c1

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -549,6 +549,19 @@ define Device/DSL274XB-F1
+  $(Device/bcm63xx)
+  IMAGES += sysupgrade.bin
+  DEVICE_TITLE := D-Link DSL-2750U rev C1
+  DEVICE_DTS := dsl-2750u-c1
```
Remove this line. DTS file has to be renamed: bcm6328-d-link-dsl-2750u-c1.dts

adrianschmutzler commented on target/linux/brcm63xx/image/bcm63xx.mk:
```diff
@@ -549,6 +549,19 @@ define Device/DSL274XB-F1
+  CFE_CHIP_ID := 6328
+  FLASH_MB := 8
+  DEVICE_PACKAGES := \
+    $(USB2_PACKAGES) $(B43_PACKAGES)
```
This will match in one line.

adrianschmutzler commented on target/linux/brcm63xx/dts/dsl-2750u-c1.dts:
```diff
@@ -0,0 +1,121 @@
+};
+
+&uart0 {
+	status = "ok";
```
"okay" should be preferred here.

adrianschmutzler commented on target/linux/brcm63xx/patches-4.14/599-board_dsl-2750u-c1.patch:
```diff
@@ -0,0 +1,68 @@
+--- a/arch/mips/bcm63xx/boards/board_bcm963xx.c
++++ b/arch/mips/bcm63xx/boards/board_bcm963xx.c
```
Please also provide patch for 4.19.

adrianschmutzler commented on target/linux/brcm63xx/base-files/lib/brcm63xx.sh:
```diff
@@ -138,6 +138,9 @@ brcm63xx_dt_detect() {
 		;;
+	"D-Link DSL-2750U rev C1")
+		board_name="dsl-2750u-c1"
+		;;
```
This is obsolete (file removed).

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/diag.sh:
```diff
@@ -64,6 +64,9 @@ set_state() {
 	dsl-274xb-f)
 		status_led="dsl-274xb:green:power"
 		;;
+	dsl-2750u-c1)
```
use compatible instead.

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/board.d/02_network:
```diff
@@ -84,6 +84,7 @@ bcm963281tan |\
 bcm96328avng |\
 bcm96368mvngr |\
 dsl-274xb-f |\
+dsl-2750u-c1 |\
```
use compatible instead.

adrianschmutzler commented on target/linux/brcm63xx/base-files/etc/board.d/01_leds:
```diff
@@ -35,6 +35,9 @@ dgnd3700v1_dgnd3800b)
 	ucidef_set_led_usbdev "usb1" "USB1" "DGND3700v1_3800B:green:usb-back" "1-1"
 	ucidef_set_led_usbdev "usb2" "USB2" "DGND3700v1_3800B:green:usb-front" "1-2"
 	;;
+dsl-2750u-c1)
```
use compatible instead.

adrianschmutzler commented:
Target rename has happen in the meantime as well.

Please update.

adrianschmutzler commented:
Closing, since this is quite old and has not been updated by author for quite some time either.
